### DISCUSSION TITLE
Insight on top-level code for callbacks

### DISCUSSION THREAD

**Discussion by nkjmsg** [apache/airflow#40983] [Q&A]
Hello,

We are seeing an increase in the number of requests that the DAGs are sending out to an external credential manager. 
We've advised all developers to reduce, or completely eliminate any top-level imports or code that could be costly or making calls to the external systems. However, something that I can't quite wrap my head around is the functions required for callbacks within the task decorators. For instance:

```
def task_failure_alert(context):
    # makes calls to the credential manager to get email credentials to send out alert emails


with DAG(...
):

@task(on_failure_callback=task_failure_alert)
   ...
```

Here we define the function task_failure_alert at top-level to be called by the task in case of a failure. It seems like this function gets parsed at every given parsing interval. So we are thinking of ways to move this out of the DAG file, and the only viable option that we could think of was to move the function out to a separate helper class and import it in the DAG. However, since imports are also parsed, we are afraid that it's just going to produce the same result. Importing it in the task isn't an option since it's being assigned to on_failure_callback. 

Is there a workaround for this? We store all credentials in the credential manager and they are retrieved at runtime, and we need them to send all DAG failure alerts. We are planning to eventually migrate to a different pattern to inject them at deployment, but it is not yet an implemented pattern. If I'm not understanding the concept of top-level code and import here, please let me know as well.

Thanks.

**Comment by potiuk** [comment]
Yes. The function gets parsed. But should not be executed. Read that chapter - it explains how to check if your code is top-level https://airflow.apache.org/docs/apache-airflow/stable/best-practices.html#top-level-python-code  - simply `python your_file.py` and raising exception in the place you expect to be executed during parsing will show you when it is atually executed.
